test_runner_lib/include/test_runner/TestOutputSpec.h

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2022 Trail of Bits, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <glog/logging.h>
+#include <remill/Arch/Arch.h>
+#include <test_runner/TestRunner.h>
+
+#include <any>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+
+namespace test_runner {
+
+using MemoryModifier = std::function<void(MemoryHandler &)>;
+using RegisterValue = std::variant<uint64_t, uint8_t>;
+using RegisterValueRef = std::variant<std::reference_wrapper<uint64_t>,
+                                      std::reference_wrapper<uint8_t>>;
+
+struct RegisterCondition {
+  std::string register_name;
+  RegisterValue enforced_value;
+};
+
+template <typename T>
+concept State = std::is_base_of_v<ArchState, T>;
+
+template <State S>
+class TestOutputSpec {
+ public:
+  uint64_t addr;
+  std::string target_bytes;
+
+ private:
+  using RegisterAccessorMap =
+      std::unordered_map<std::string, std::function<RegisterValueRef(S &)>>;
+  using RegisterConditionList = std::vector<RegisterCondition>;
+  using MemoryConditionList = std::vector<MemoryModifier>;
+
+  remill::Instruction::Category expected_category;
+  RegisterConditionList register_preconditions;
+  RegisterConditionList register_postconditions;
+  MemoryConditionList initial_memory_conditions;
+  MemoryConditionList expected_memory_conditions;
+  RegisterAccessorMap reg_to_accessor;
+
+  template <typename T>
+  void ApplyCondition(S &state, const std::string &reg, T value) const {
+    auto accessor = reg_to_accessor.find(reg);
+    if (accessor != reg_to_accessor.end()) {
+      auto wrapper = accessor->second(state);
+      if (auto underlying = std::get_if<std::reference_wrapper<T>>(&wrapper)) {
+        underlying->get() = value;
+      } else {
+        throw std::runtime_error(
+            std::string("Reg value " + reg + " has incorrect type"));
+      }
+    } else {
+      throw std::runtime_error(std::string("Unknown reg: ") + reg);
+    }
+  }
+
+  template <typename T>
+  void CheckCondition(S &state, const std::string &reg, T value) const {
```

Comment:
And since we need the whole mechanism here again:
```
T  &get_reg( ... )
{
  if ( !accesor_present )
    throw;
  return get_if_helper();
}
 
void CheckCond( ... ) { CHECK_EQ( actual, helper( ... ) ) }
void Apply( ... ) { get_reg( ... ) = value; } 
```